File: Part 2 - Regression/Random Forest Regression/random_forest_regression.r
# Random Forest Regression


#importando dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[,2:3]

#Dividir data set entre conjunto de entranamiento y testing
#library(caTools)
#set.seed(123) #Seleccionando semilla
#sample.split(Datos a separar, porcentaje para entrenar)
#split = sample.split(dataset$Purchased,SplitRatio = 0.8)
#Separando datos de entrenamiento y testing
#training_set = subset(dataset,split==TRUE)
#testing_set = subset(dataset,split==FALSE)

#Escalado de valores
#training_set[,2:3] = scale(training_set[,2:3])#[Todas las filas, de 2 a 3]
#testing_set[,2:3] = scale(testing_set[,2:3])

#Ajustar el modelo de regresion con el conjunto de datos
install.packages("randomForest")
library(randomForest)
set.seed(1234)
#dataframe = con []
#vector = con $
regression = randomForest(x = dataset[1], 
                          y = dataset$Salary,
                          ntree = 550)

#Prediccion de nuevos resultado con Random Forest Regression
y_pred = predict(regression, newdata=data.frame(Level = 6.5))



#Visualizacion del modelo de Random Forest Regression
library(ggplot2)
x_grid = seq(min(dataset$Level),max(dataset$Level),0.01)
ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = "red") +#colocando puntos
  geom_line(aes(x = x_grid, y = predict(regression, newdata=data.frame(Level = x_grid))),
            colour = "blue") + #Colocando recta de regresion
  ggtitle("Prediccion (Random Forest Regression)")+
  xlab("Nivel del empleado") +
  ylab("Sueldo (en $)")

File: Part 2 - Regression/Random Forest Regression/random_forest_regression.py
# -*- coding: utf-8 -*-
"""
Created on Sun Apr  9 21:00:28 2023

@author: josue
"""


# Regresion con Bosques Aleatorios

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

dataset = pd.read_csv('Position_Salaries.csv')

# matriz. Obteniendo valores del data frame [filas,columnas]
X = dataset.iloc[:, 1:2].values #Matriz 
y = dataset.iloc[:, 2].values  # vector


#Dividir data set entre conjunto de entranamiento y testing
#from sklearn.model_selection import train_test_split
#X_train,X_test,y_train,y_test=train_test_split(X,y,test_size = 0.2, random_state=0)


#Escalado de variables
"""
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
"""

# Ajustar la regresion con el dataset
from sklearn.ensemble import RandomForestRegressor
#A la hora de cortar se usa el error cuadrado medio tipicamente
#n_estimators: Numero de arboles
regression = RandomForestRegressor(n_estimators = 330, random_state = 0)
regression.fit(X,y)

# Prediccion de nuestros modelos
y_pred = regression.predict([[6.5]])


# Visualización de los resultados del Bosque Aleatorio
X_grid = np.arange(min(X),max(X),0.01)
X_grid = X_grid.reshape(len(X_grid),1)
plt.scatter(X, y, color = "red")
plt.plot(X_grid, regression.predict(X_grid), color = "blue")
plt.title("Modelo de Bosques Aleatorios para Regresión")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo (en $)")
plt.show()


